File: modules/feed-parser/ios/FeedParserModule.swift
import ExpoModulesCore
import Foundation

public class FeedParserModule: Module {
  public func definition() -> ModuleDefinition {
    Name("FeedParser")

    AsyncFunction("parseFeed") { (feedUrl: String) async throws -> [String: Any] in
      let xml = try await self.fetchFeed(url: feedUrl)
      return self.parseFeedXml(xml)
    }
  }

  private func fetchFeed(url: String) async throws -> String {
    guard let url = URL(string: url) else {
      throw NSError(domain: "FeedParser", code: 1, userInfo: [NSLocalizedDescriptionKey: "Invalid URL"])
    }
    let (data, _) = try await URLSession.shared.data(from: url)
    guard let xml = String(data: data, encoding: .utf8) else {
      throw NSError(domain: "FeedParser", code: 2, userInfo: [NSLocalizedDescriptionKey: "Failed to decode response"])
    }
    return xml
  }

  private func parseFeedXml(_ xml: String) -> [String: Any] {
    guard let data = xml.data(using: .utf8) else {
      return ["title": "", "imageUrl": "", "episodes": []]
    }
    let delegate = FeedParserDelegate()
    let parser = XMLParser(data: data)
    parser.delegate = delegate
    parser.parse()
    return [
      "title": delegate.channelTitle,
      "description": delegate.channelDescription,
      "author": delegate.channelAuthor,
      "imageUrl": delegate.channelImageUrl,
      "episodes": delegate.episodes,
    ]
  }
}

private class FeedParserDelegate: NSObject, XMLParserDelegate {
  var channelTitle = ""
  var channelDescription = ""
  var channelAuthor = ""
  var channelImageUrl = ""
  var episodes: [[String: Any]] = []

  private var inChannel = false
  private var inItem = false
  private var inImage = false
  private var currentElement = ""
  private var textBuffer = ""

  private var itemTitle = ""
  private var itemDescription = ""
  private var itemPubDate = ""
  private var itemDuration = ""
  private var itemAudioUrl = ""
  private var itemFileSize: Int64 = 0
  private var itemImageUrl = ""
  private var itemCount = 0

  private static let dateFormatters: [DateFormatter] = {
    let formats = [
      "EEE, dd MMM yyyy HH:mm:ss Z",
      "EEE, dd MMM yyyy HH:mm:ss zzz",
      "yyyy-MM-dd'T'HH:mm:ssZ",
      "yyyy-MM-dd'T'HH:mm:ss.SSSZ",
      "yyyy-MM-dd'T'HH:mm:ssXXX",
    ]
    return formats.map { format in
      let formatter = DateFormatter()
      formatter.dateFormat = format
      formatter.locale = Locale(identifier: "en_US_POSIX")
      return formatter
    }
  }()

  func parser(_ parser: XMLParser, didStartElement elementName: String,
              namespaceURI: String?, qualifiedName qName: String?,
              attributes: [String: String] = [:]) {
    currentElement = elementName
    textBuffer = ""

    switch elementName {
    case "channel":
      inChannel = true
    case "item" where inChannel:
      inItem = true
      itemTitle = ""
      itemDescription = ""
      itemPubDate = ""
      itemDuration = ""
      itemAudioUrl = ""
      itemFileSize = 0
      itemImageUrl = ""
    case "image" where inChannel && !inItem:
      inImage = true
    case "enclosure" where inItem:
      itemAudioUrl = attributes["url"] ?? ""
      itemFileSize = Int64(attributes["length"] ?? "") ?? 0
    case "itunes:image":
      let href = attributes["href"] ?? ""
      if inItem {
        itemImageUrl = href
      } else if inChannel && !inItem && channelImageUrl.isEmpty {
        channelImageUrl = href
      }
    default:
      break
    }
  }

  func parser(_ parser: XMLParser, foundCharacters string: String) {
    textBuffer += string
  }

  func parser(_ parser: XMLParser, didEndElement elementName: String,
              namespaceURI: String?, qualifiedName qName: String?) {
    let text = textBuffer.trimmingCharacters(in: .whitespacesAndNewlines)

    if elementName == "item" && inItem {
      if !itemAudioUrl.isEmpty && itemCount < 50 {
        episodes.append([
          "title": itemTitle.isEmpty ? "Untitled" : itemTitle,
          "description": stripHtml(itemDescription),
          "audioUrl": itemAudioUrl,
          "imageUrl": itemImageUrl.isEmpty ? channelImageUrl : itemImageUrl,
          "publishedAt": parseDate(itemPubDate),
          "duration": parseDuration(itemDuration),
          "fileSize": itemFileSize,
        ])
        itemCount += 1
      }
      inItem = false
      return
    }

    if elementName == "image" { inImage = false; return }
    if elementName == "channel" { inChannel = false; return }

    if inItem {
      switch elementName {
      case "title": itemTitle = text
      case "description": itemDescription = text
      case "itunes:summary" where itemDescription.isEmpty: itemDescription = text
      case "pubDate": itemPubDate = text
      case "itunes:duration": itemDuration = text
      default: break
      }
    } else if inChannel {
      if elementName == "title" && channelTitle.isEmpty {
        channelTitle = text
      } else if elementName == "description" && channelDescription.isEmpty {
        channelDescription = stripHtml(text)
      } else if elementName == "itunes:author" && channelAuthor.isEmpty {
        channelAuthor = text
      } else if elementName == "managingEditor" && channelAuthor.isEmpty {
        channelAuthor = text
      } else if elementName == "url" && inImage && channelImageUrl.isEmpty {
        channelImageUrl = text
      }
    }
  }

  private func parseDate(_ dateStr: String) -> Double {
    if dateStr.isEmpty { return Date().timeIntervalSince1970 * 1000 }
    for formatter in Self.dateFormatters {
      if let date = formatter.date(from: dateStr) {
        return date.timeIntervalSince1970 * 1000
      }
    }
    return Date().timeIntervalSince1970 * 1000
  }

  private func parseDuration(_ duration: String) -> Double {
    if duration.isEmpty { return 0 }
    if let num = Double(duration) { return num }
    let parts = duration.split(separator: ":").compactMap { Int($0) }
    switch parts.count {
    case 3: return Double(parts[0] * 3600 + parts[1] * 60 + parts[2])
    case 2: return Double(parts[0] * 60 + parts[1])
    case 1: return Double(parts[0])
    default: return 0
    }
  }

  private func stripHtml(_ html: String) -> String {
    var result = html
    result = result.replacingOccurrences(of: "<br\\s*/?>", with: "\n", options: .regularExpression)
    result = result.replacingOccurrences(of: "<p\\s*>", with: "\n", options: .regularExpression)
    result = result.replacingOccurrences(of: "</p>", with: "\n")
    result = result.replacingOccurrences(of: "<[^>]+>", with: "", options: .regularExpression)
    result = result.replacingOccurrences(of: "&amp;", with: "&")
    result = result.replacingOccurrences(of: "&lt;", with: "<")
    result = result.replacingOccurrences(of: "&gt;", with: ">")
    result = result.replacingOccurrences(of: "&quot;", with: "\"")
    result = result.replacingOccurrences(of: "&#39;", with: "'")
    result = result.replacingOccurrences(of: "&apos;", with: "'")
    result = result.replacingOccurrences(of: "&#x27;", with: "'")
    result = result.replacingOccurrences(of: "&nbsp;", with: " ")
    result = result.replacingOccurrences(of: "\n{3,}", with: "\n\n", options: .regularExpression)
    return result.trimmingCharacters(in: .whitespacesAndNewlines)
  }
}

File: modules/feed-parser/android/src/main/java/expo/modules/feedparser/FeedParserModule.kt
package expo.modules.feedparser

import android.util.Xml
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserException
import java.io.StringReader
import java.text.SimpleDateFormat
import java.util.Locale
import java.util.concurrent.TimeUnit

class FeedParserModule : Module() {
  private val client = OkHttpClient.Builder()
    .connectTimeout(30, TimeUnit.SECONDS)
    .readTimeout(30, TimeUnit.SECONDS)
    .followRedirects(true)
    .followSslRedirects(true)
    .build()

  private val dateFormats = listOf(
    SimpleDateFormat("EEE, dd MMM yyyy HH:mm:ss Z", Locale.US),
    SimpleDateFormat("EEE, dd MMM yyyy HH:mm:ss zzz", Locale.US),
    SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ", Locale.US),
    SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSZ", Locale.US),
    SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssXXX", Locale.US),
  )

  override fun definition() = ModuleDefinition {
    Name("FeedParser")

    AsyncFunction("parseFeed") Coroutine { feedUrl: String ->
      withContext(Dispatchers.IO) {
        val xml = fetchFeed(feedUrl)
        parseFeedXml(xml)
      }
    }
  }

  private fun fetchFeed(url: String): String {
    val request = Request.Builder()
      .url(url)
      .header("User-Agent", "Pegycast/1.0")
      .header("Accept", "application/rss+xml, application/xml, text/xml, */*")
      .build()
    val response = client.newCall(request).execute()
    return response.body?.string() ?: throw Exception("Empty response from $url")
  }

  private fun parseFeedXml(xml: String): Map<String, Any?> {
    val parser = Xml.newPullParser()
    parser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, false)
    parser.setInput(StringReader(xml))

    var channelTitle = ""
    var channelDescription = ""
    var channelAuthor = ""
    var channelImageUrl = ""
    val episodes = mutableListOf<Map<String, Any?>>()

    var inChannel = false
    var inItem = false
    var inImage = false
    var currentTag = ""
    var textBuffer = StringBuilder()

    // Current item fields
    var itemTitle = ""
    var itemDescription = ""
    var itemPubDate = ""
    var itemDuration = ""
    var itemAudioUrl = ""
    var itemFileSize = 0L
    var itemImageUrl = ""

    var itemCount = 0

    while (parser.eventType != XmlPullParser.END_DOCUMENT) {
      when (parser.eventType) {
        XmlPullParser.START_TAG -> {
          val tag = parser.name
          textBuffer = StringBuilder()

          when {
            tag == "channel" -> inChannel = true
            tag == "item" && inChannel -> {
              inItem = true
              itemTitle = ""
              itemDescription = ""
              itemPubDate = ""
              itemDuration = ""
              itemAudioUrl = ""
              itemFileSize = 0L
              itemImageUrl = ""
            }
            tag == "image" && inChannel && !inItem -> inImage = true
            tag == "enclosure" && inItem -> {
              itemAudioUrl = parser.getAttributeValue(null, "url") ?: ""
              val length = parser.getAttributeValue(null, "length")
              itemFileSize = length?.toLongOrNull() ?: 0L
            }
            tag == "itunes:image" -> {
              val href = parser.getAttributeValue(null, "href") ?: ""
              if (inItem) {
                itemImageUrl = href
              } else if (inChannel && !inItem) {
                if (channelImageUrl.isEmpty()) channelImageUrl = href
              }
            }
          }
          currentTag = tag
        }

        XmlPullParser.TEXT -> {
          textBuffer.append(parser.text)
        }

        XmlPullParser.END_TAG -> {
          val tag = parser.name
          val text = textBuffer.toString().trim()

          when {
            tag == "item" && inItem -> {
              if (itemAudioUrl.isNotEmpty() && itemCount < 50) {
                episodes.add(mapOf(
                  "title" to (itemTitle.ifEmpty { "Untitled" }),
                  "description" to stripHtml(itemDescription),
                  "audioUrl" to itemAudioUrl,
                  "imageUrl" to itemImageUrl.ifEmpty { channelImageUrl },
                  "publishedAt" to parseDate(itemPubDate),
                  "duration" to parseDuration(itemDuration),
                  "fileSize" to itemFileSize,
                ))
                itemCount++
              }
              inItem = false
            }
            tag == "image" -> inImage = false
            tag == "channel" -> inChannel = false
          }

          if (inItem) {
            when (currentTag) {
              "title" -> if (tag == "title") itemTitle = text
              "description" -> if (tag == "description") itemDescription = text
              "itunes:summary" -> if (tag == "itunes:summary" && itemDescription.isEmpty()) itemDescription = text
              "pubDate" -> if (tag == "pubDate") itemPubDate = text
              "itunes:duration" -> if (tag == "itunes:duration") itemDuration = text
            }
          } else if (inChannel) {
            when {
              tag == "title" && channelTitle.isEmpty() -> channelTitle = text
              tag == "description" && channelDescription.isEmpty() -> channelDescription = stripHtml(text)
              tag == "itunes:author" && channelAuthor.isEmpty() -> channelAuthor = text
              tag == "managingEditor" && channelAuthor.isEmpty() -> channelAuthor = text
              tag == "url" && inImage && channelImageUrl.isEmpty() -> channelImageUrl = text
            }
          }
        }
      }
      parser.next()
    }

    return mapOf(
      "title" to channelTitle,
      "description" to channelDescription,
      "author" to channelAuthor,
      "imageUrl" to channelImageUrl,
      "episodes" to episodes,
    )
  }

  private fun parseDate(dateStr: String): Double {
    if (dateStr.isEmpty()) return System.currentTimeMillis().toDouble()
    for (format in dateFormats) {
      try {
        return format.parse(dateStr)?.time?.toDouble() ?: continue
      } catch (_: Exception) {
        continue
      }
    }
    return System.currentTimeMillis().toDouble()
  }

  private fun parseDuration(duration: String): Double {
    if (duration.isEmpty()) return 0.0
    duration.toDoubleOrNull()?.let { return it }
    val parts = duration.split(":").mapNotNull { it.toIntOrNull() }
    return when (parts.size) {
      3 -> (parts[0] * 3600 + parts[1] * 60 + parts[2]).toDouble()
      2 -> (parts[0] * 60 + parts[1]).toDouble()
      1 -> parts[0].toDouble()
      else -> 0.0
    }
  }

  private fun stripHtml(html: String): String {
    return html
      .replace(Regex("<br\\s*/?>"), "\n")
      .replace(Regex("<p\\s*>"), "\n")
      .replace(Regex("</p>"), "\n")
      .replace(Regex("<[^>]+>"), "")
      .replace("&amp;", "&")
      .replace("&lt;", "<")
      .replace("&gt;", ">")
      .replace("&quot;", "\"")
      .replace("&#39;", "'")
      .replace("&apos;", "'")
      .replace("&#x27;", "'")
      .replace("&nbsp;", " ")
      .replace(Regex("\n{3,}"), "\n\n")
      .trim()
  }
}